File: frontend/src/types/graph.ts
export interface GraphNode {
  id: string;
  label: string;
  layer: number;
  layer_name: string;
  style_class: string;
  fill_color: string;
  stroke_color: string;
  mastery: number | null;
  card_count: number;
}

export interface GraphEdge {
  source: string;
  target: string;
  type: string;
}

export interface KnowledgeGraph {
  nodes: GraphNode[];
  edges: GraphEdge[];
  layers: Record<number, string>;
}

export type GraphNodeType = 'concept' | 'subtopic' | 'card';

export interface NavItem {
  id: string;
  type: GraphNodeType;
  name: string;
}

export interface Subtopic {
  id: string;
  name: string;
  card_count: number;
}

export interface SubtreeCardBreakdownItem {
  concept: string;
  count: number;
  is_prerequisite: boolean;
}

export interface SubtreeCardDistribution {
  node_id: string;
  total: number;
  breakdown: SubtreeCardBreakdownItem[];
}

File: backend/app/models/graph.py
from pydantic import BaseModel


class GraphNode(BaseModel):
    id: str
    label: str
    layer: int
    layer_name: str
    style_class: str
    fill_color: str
    stroke_color: str
    mastery: float | None = None
    card_count: int = 0


class GraphEdge(BaseModel):
    source: str
    target: str
    type: str = "prerequisite"


class KnowledgeGraph(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    layers: dict[int, str]


class SubtopicSummary(BaseModel):
    id: str
    name: str
    card_count: int


class SubtreeCardBreakdownItem(BaseModel):
    concept: str
    count: int
    is_prerequisite: bool


class SubtreeCardDistribution(BaseModel):
    node_id: str
    total: int
    breakdown: list[SubtreeCardBreakdownItem]
